<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Juan Pablo Rodríguez Alonso, Ana Esmeralda Rodríguez Rodríguez.

**Fecha:** 11 de junio del 2021.

**Expediente** : 715180, 709288.

**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: [Link con el enlace del repositorio del creador del proyecto](https://github.com/JuanPrdz/Tarea4_JPRodriguez_ARodriguez)

# Tarea 4: Clase 7



## Ejercicio 1
Como ejemplo simple de una simulación de Monte Carlo, considere calcular la probabilidad de una suma particular del lanzamiento de tres dados (cada dado tiene valores del uno al seis). Además cada dado tiene las siguientes carácterísticas: el primer dado no está cargado (distribución uniforme todos son equiprobables); el segundo y tercer dado están cargados basados en una distribución binomial con parámetros (`n=5, p=0.5` y `n=5, p=0.2`). Calcule la probabilidad de que la suma resultante sea 7, 14 o 18. 

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
import scipy.stats as st
import seaborn as sns
sns.set()

#### Ejercicio 1: Estudiante 1 Juan Pablo Rodríguez Alonso

In [3]:
# Código de solución estudiante 1 Juan Pablo Rodríguez

N = 1000000
np.random.seed(22)
d_no_cargado = np.random.randint(1,7, N)
d_cargado_1 = st.binom(n=5, p=0.5, loc=1).rvs(size=N)
d_cargado_2 = st.binom(n=5, p=0.2, loc=1).rvs(size=N)
suma = d_no_cargado + d_cargado_1 + d_cargado_2

resultantes = [7, 14, 18]
[len(suma[suma == i])/len(suma) for i in resultantes]

[0.127451, 0.013755, 1e-06]

##### La probabilidad de que la suma de los dados sea:
* 7 es de 0.1274
* 14 es de 0.01375
* 18 es de  1e-06

#### Ejercicio 1: Estudiante 2 Ana Esmeralda Rodríguez

In [4]:
# Código de solución estudiante 2 Ana Esmeralda Rodríguez

n = 100000
# DADO 1 (NO CARGADO)
dado1 = np.random.randint(1,7,n)
# DADO 2 (CARGADO n=5, p=0.5)
dado2 = st.binom(n=5, p=0.5).rvs(size=n)
# DADO 3 (CARGADO  n=5, p=0.2)
dado3 = st.binom(n=5, p=0.2).rvs(size=n)

x = dado1 + dado2 + dado3
suma_7 = (sum([(x == 7).sum()]))/n 
suma_14 = (sum([(x == 14).sum()]))/n
suma_18 = (sum([(x == 18).sum()]))/n
suma_7, suma_14, suma_18

(0.16296, 0.00033, 0.0)

In [5]:
print("La probabilidad de que la suma de los 3 dados sea 7 es de:", suma_7)
print("La probabilidad de que la suma de los 3 dados sea 14 es de:", suma_14)
print("La probabilidad de que la suma de los 3 dados sea 18 es de:", suma_18)

La probabilidad de que la suma de los 3 dados sea 7 es de: 0.16296
La probabilidad de que la suma de los 3 dados sea 14 es de: 0.00033
La probabilidad de que la suma de los 3 dados sea 18 es de: 0.0


# 2 Ejercicio de aplicación- Cafetería Central

Premisas para la simulación:
- Negocio de alimentos que vende bebidas y alimentos.
- Negocio dentro del ITESO.
- Negocio en cafetería central.
- Tipo de clientes (hombres y mujeres).
- Rentabilidad del 60%.

## Objetivo
Realizar una simulación estimado el tiempo medio que se tardaran los clientes en ser atendidos entre el horario de 6:30 a 1 pm. Además saber el consumo. 
**Analizar supuestos y limitantes**
## Supuestos en simulación
Clasificación de clientes: 
- Mujer  = 1 $\longrightarrow$ aleatorio < 0.5
- Hombre = 0 $\longrightarrow$ aleatorio $\geq$ 0.5.

Condiciones iniciales:
- Todas las distrubuciones de probabilidad se supondrán uniformes.
- Tiempo de simulación: 6:30 am - 1:30pm $\longrightarrow$ T = 7 horas = 25200 seg.
- Tiempo de llegada hasta ser atendido: Min=5seg, Max=30seg.
- Tiempo que tardan los clientes en ser atendidos:
    - Mujer: Min = 1 min= 60seg, Max = 5 min = 300 seg
    - Hombre: Min = 40 seg, Max = 2 min= 120 seg
- Consumo según el tipo de cliente:
    - Mujer: Min = 30 pesos, Max = 100 pesos
    - Hombre: Min = 20 pesos, Max = 80 pesos

Responder las siguientes preguntas basados en los datos del problema:
1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.
2. ¿Cuál fue el consumo promedio de los hombres y mujeres?
3. ¿Cuál fue el número de personas atendidas por día?
4. ¿Cuál fue el tiempo de atención promedio?
5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?

#### Ejercicio 2: Estudiante 1 Juan Pablo Rodríguez Alonso

In [4]:
# Código de solución estudiante 1 Juan Pablo Rodríguez


################## Datos del problema
d = 5
T =25200
T_at_min = 5; T_at_max = 30
T_mujer_min =60; T_mujer_max = 300
T_hombre_min = 40; T_hombre_max = 120
C_mujer_min = 30; C_mujer_max = 100
C_hombre_min = 20; C_hombre_max = 80

def hombre():
    """"
    Función que da un arreglo de un hombre, en la primera posición el tiempo que 
    va a ocupar, en la segunda posición su gasto en consumo"""
    
    return [np.random.uniform(T_hombre_min, T_hombre_max) + 
            np.random.uniform(T_at_min, T_at_max), np.random.uniform(C_hombre_min, C_hombre_max)]
def mujer():
    """"
    Función que da un arreglo de una mujer, en la primera posición el tiempo que 
    va a ocupar, en la segunda posición su gasto en consumo"""
    
    return [np.random.uniform(T_mujer_min, T_mujer_max) + 
            np.random.uniform(T_at_min, T_at_max), np.random.uniform(C_mujer_min, C_mujer_max)]

m_cons = []; h_cons = []
c_c = 0; clientes = []

#def cafe()
for i in range(d):
    t_total = 0
    c_c = 0
    while t_total<T:
        sexo = np.random.rand()
        if sexo < 0.5:
            des = mujer()
            m_cons.append(des[1])
        else:
            des = hombre()
            h_cons.append(des[1])
        c_c += 1
        t_total = t_total + des[0]
    clientes.append(c_c)
        
print("1. Los gastos de hombres en los 5 días de trabajo fueron: ", np.array(h_cons).sum(), " y los de las mujeres: ", np.array(m_cons).sum())
print("2. El consumo promedio de los hombre fue: ", np.array(h_cons).mean(), " y el de las mujeres: ", np.array(m_cons).mean())
print("3. El número de personas atendidas por día fue de (cada día): ", clientes)
print("4. El tiempo promedio de atención fue de: ", (np.array(t_total)/np.array(clientes)).mean(), " por día.")
print("5. La ganancia promedio con su rentabilidad del 60% en los 5 días de trabajo fue: ", np.array(h_cons).sum()*.6+np.array(m_cons).sum()*.6)

1. Los gastos de hombres en los 5 días de trabajo fueron:  24181.151070486034  y los de las mujeres:  26682.341384487332
2. El consumo promedio de los hombre fue:  52.114549720875075  y el de las mujeres:  64.60615347333494
3. El número de personas atendidas por día fue de (cada día):  [181, 176, 173, 169, 178]
4. El tiempo promedio de atención fue de:  145.113211468658  por día.
5. La ganancia promedio con su rentabilidad del 60% en los 5 días de trabajo fue:  30518.095472984016


#### Ejercicio 2: Estudiante 2 Ana Esmeralda Rodríguez

In [8]:
# Código de solución estudiante 2 Ana Esmeralda Rodríguez

################## Datos del problema
d = 5
T =25200
T_at_min = 5; T_at_max = 30
T_mujer_min =60; T_mujer_max = 300
T_hombre_min = 40; T_hombre_max = 120
C_mujer_min = 30; C_mujer_max = 100
C_hombre_min = 20; C_hombre_max = 80

# SIMULACIÓN TIEMPO DE ESPERA ALEATORIO
t_general = np.random.randint(T_at_min,T_at_max+1)
t_mujer = np.random.randint(T_mujer_min,T_mujer_max+1)
t_hombre = np.random.randint(T_hombre_min,T_hombre_max+1)
# SIMULACIÓN CONSUMO ALEATORIO
c_mujer = np.random.randint(C_mujer_min,C_mujer_max+1)
c_hombre = np.random.randint(C_hombre_min,C_hombre_max+1)
# SIMULACIÓN 200 CLIENTES ALEATORIOS (PARA EL GÉNERO)
clientes = np.random.rand(200)

In [9]:
T_atencion = []
ganancia = []
rentabilidad = []
# Simulación de 1 día 
def sim1dia(clientes,T):
    i = 0
    t_atencion = 0 
    cantidad_mujeres = 0 
    cantidad_hombres = 0 
    consumo_total_hombres = 0  
    consumo_total_mujeres= 0  
    while t_atencion <= T:
        t_llegada = t_general
        t_atencion += t_llegada
        if clientes[i] < 0.5:
            t_persona = t_mujer
            t_atencion += t_persona
            cantidad_mujeres += 1
            consumo_total_mujeres += c_mujer
        else:
            t_persona = t_hombre
            t_atencion += t_persona
            cantidad_hombres += 1
            consumo_total_hombres += c_hombre
        
        i += 1
        T_atencion.append(t_persona)
        ganancia.append((consumo_total_hombres+consumo_total_mujeres))
        rentabilidad.append((consumo_total_hombres+consumo_total_mujeres)*0.6)
    return np.array([cantidad_mujeres,cantidad_hombres,consumo_total_hombres,consumo_total_mujeres,
                     i,(consumo_total_hombres+consumo_total_mujeres)*0.6])
def dias_5(n:'Número de días'):
    return [sim1dia(np.random.rand(200), T) for i in range(n)]

sim5dias = dias_5(d)

In [22]:
print('1. Los gastos de los hombres en 5 días de trabajo fue de:', sum(list(map(lambda x: x[2],sim5dias))))
print('   y el de las mujeres de:', sum(list(map(lambda x: x[3],sim5dias))))
print('2. El consumo promedio de los hombres fue de:', sum(list(map(lambda x: x[2],sim5dias))) / (sum(list(map(lambda x: x[1],sim5dias)))))
print('   y el de las mujeres de:', sum(list(map(lambda x: x[3],sim5dias)))/ (sum(list(map(lambda x: x[0],sim5dias)))))
print('3. El número de personas atendidas en un dia es de:', sum(list(map(lambda x: x[0] + x[1] ,sim5dias))) / d)
print('4. El tiempo promedio de espra fue de:', np.mean(T_atencion))
print('5. La ganancia promedio de la cafetería en 5 días fue de:', np.mean(ganancia))
print('   La rentabilidad promedio de la cafetería en 5 días fue de:', np.mean(rentabilidad))


1. Los gastos de los hombres en 5 días de trabajo fue de: 16896.0
   y el de las mujeres de: 26599.0
2. El consumo promedio de los hombres fue de: 44.0
   y el de las mujeres de: 67.0
3. El número de personas atendidas en un dia es de: 156.2
4. El tiempo promedio de espra fue de: 153.94878361075544
5. La ganancia promedio de la cafetería en 5 días fue de: 4342.623559539053
   La rentabilidad promedio de la cafetería en 5 días fue de: 2605.5741357234315
